src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -85,71 +91,117 @@ class TrimPrimers
     val maxPrimerLength = detector.getAll.map(_.longestPrimerLength).max
 
+    // Main processing loop
     val iterator = queryNameOrderIterator(in)
+    val trimProgress = new ProgressLogger(this.logger, "Trimmed")
     while (iterator.hasNext) {
       val reads = nextTemplate(iterator)
-      val rec1 = reads.find(r => r.getReadPairedFlag && r.getFirstOfPairFlag  && !r.isSecondaryOrSupplementary)
-      val rec2 = reads.find(r => r.getReadPairedFlag && r.getSecondOfPairFlag && !r.isSecondaryOrSupplementary)
-
-      (rec1, rec2) match {
-        case (Some(r1), Some(r2)) =>
-          // FR mapped pairs get the full treatment
-          if (!r1.getReadUnmappedFlag && !r2.getReadUnmappedFlag &&
-            r1.getReferenceName == r2.getReferenceName &&
-            SamPairUtil.getPairOrientation(r1) == PairOrientation.FR) {
-
-            val (left, right) = if (r1.getReadNegativeStrandFlag) (r2, r1) else (r1, r2)
-            val (start, end ) = (left.getUnclippedStart, right.getUnclippedEnd)
-            val insert = new Interval(left.getContig, start, end)
-            detector.getOverlaps(insert).find(amp => abs(amp.leftStart - start) <= slop && abs(amp.rightEnd - end) <= slop) match {
-              case Some(amplicon) =>
-                val leftClip  = amplicon.leftPrimerLength
-                val rightClip = amplicon.rightPrimerLength
-                reads.foreach { rec =>
-                  val toClip = if (rec.getFirstOfPairFlag == left.getFirstOfPairFlag) leftClip else rightClip
-                  SamRecordClipper.clip5PrimeEndOfRead(rec, toClip, mode)
-                }
-              case None =>
-                reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
-            }
-
-            clipFullyOverlappedFrReads(r1, r2)
-          }
-          // Pairs without both reads mapped in FR orientation are just maximally clipped
-          else {
-            reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
-          }
+      trimReadsForTemplate(detector, maxPrimerLength, reads)
+      reads.foreach(write)
+      reads.foreach(trimProgress.record)
+    }
 
-          SamPairUtil.setMateInfo(r1, r2, true)
-          reads.filter(_.getSupplementaryAlignmentFlag).foreach { rec =>
-            val mate = if (rec.getFirstOfPairFlag) r2 else r1
-            SamPairUtil.setMateInformationOnSupplementalAlignment(rec, mate, true);
-          }
-        case _ =>
-          // Just trim each read independently
-          reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
-      }
+    // If we had a reference and re-sorted above, reset the NM/UQ/MD tags as we push to the final output
+    (sorter, ref) match {
+      case (Some(recs: SortingCollection[SAMRecord]), Some(path)) =>
+        val walker = new ReferenceSequenceFileWalker(path.toFile)
+        val progress = new ProgressLogger(this.logger, "Written")
```

Comment:
`iterator()` calls `doneAdding` internally.  You only need to call it manually if you want to sort the last chunk and write it out immediately because you're not going to use the iterator for a while.